In [2]:
import duckdb
con = duckdb.connect("../data/football.duckdb", read_only=True)

In [3]:
con.sql("""
    DESCRIBE raw.players
""").show()

┌──────────────────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│             column_name              │ column_type │  null   │   key   │ default │  extra  │
│               varchar                │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ player_id                            │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ first_name                           │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ last_name                            │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ name                                 │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ last_season                          │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ current_club_id                      │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ player_code                          │ VARCHAR  

In [4]:
con.sql("""
    SELECT * FROM raw.players
    LIMIT 10
""").show()

┌───────────┬────────────┬──────────────┬─────────────────────┬─────────────┬─────────────────┬─────────────────────┬────────────────────┬──────────────────┬────────────────────────┬─────────────────────┬────────────────────┬────────────┬─────────┬──────────────┬──────────────────────────┬──────────────────────┬─────────────────────────────────────────────────────────────────────────────────┬────────────────────┬─────────────────────┬──────────────────────────┬────────────────────────────────────────────────────────────────────────┬──────────────────────────────────────┬─────────────────────────────────────────────────────────┬─────────────────────┬─────────────────────────────┐
│ player_id │ first_name │  last_name   │        name         │ last_season │ current_club_id │     player_code     │  country_of_birth  │  city_of_birth   │ country_of_citizenship │    date_of_birth    │    sub_position    │  position  │  foot   │ height_in_cm │ contract_expiration_date │      agent_name      

In [5]:
con.sql("""
    DESCRIBE raw.competitions
""").show()

┌──────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│     column_name      │ column_type │  null   │   key   │ default │  extra  │
│       varchar        │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ competition_id       │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ competition_code     │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ name                 │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ sub_type             │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ type                 │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ country_id           │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ country_name         │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ domestic_league_code │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ confederation        │ VARCHAR     │ YES     │ NUL

In [5]:
con.sql("""
SELECT competition_id, country_name, name
FROM raw.competitions
ORDER BY country_name
""").show(max_rows=100)


┌────────────────┬────────────────┬─────────────────────────────────────────┐
│ competition_id │  country_name  │                  name                   │
│    varchar     │    varchar     │                 varchar                 │
├────────────────┼────────────────┼─────────────────────────────────────────┤
│ ARG1           │ Argentina      │ torneo-apertura                         │
│ AUS1           │ Australia      │ a-league-men                            │
│ A1             │ Austria        │ bundesliga                              │
│ POBE           │ Belgium        │ jupiler-pro-league-europe-play-offs     │
│ BE1            │ Belgium        │ jupiler-pro-league                      │
│ BESC           │ Belgium        │ volkswagen-supercup                     │
│ BPO4           │ Belgium        │ jupiler-pro-league-relegation-play-offs │
│ EJPL           │ Belgium        │ jupiler-pro-league-playoff-i            │
│ BRA1           │ Brazil         │ campeonato-brasileiro-serie-

In [8]:
con.sql("""
SELECT competition_id, country_name, name
FROM raw.competitions
WHERE LOWER(country_name) = 'poland'
""").show(max_rows=100)

┌────────────────┬──────────────┬────────────────────┐
│ competition_id │ country_name │        name        │
│    varchar     │   varchar    │      varchar       │
├────────────────┼──────────────┼────────────────────┤
│ PL1            │ Poland       │ pko-bp-ekstraklasa │
└────────────────┴──────────────┴────────────────────┘



In [14]:
con.sql("""
DESCRIBE raw.games
""").show()

┌────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│      column_name       │ column_type │  null   │   key   │ default │  extra  │
│        varchar         │   varchar   │ varchar │ varchar │ varchar │ varchar │
├────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ game_id                │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ competition_id         │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ season                 │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ round                  │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ date                   │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ home_club_id           │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ away_club_id           │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ home_club_goals        │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ away_club_goals        │ B

In [20]:
con.sql("""
SELECT season,
COUNT (*) AS total_games
FROM raw.games
WHERE competition_id = 'PL1'
GROUP BY season
ORDER BY season DESC
""").show(max_rows=100)

┌────────┬─────────────┐
│ season │ total_games │
│ int64  │    int64    │
├────────┼─────────────┤
│   2025 │         306 │
│   2024 │         306 │
└────────┴─────────────┘



In [21]:
con.sql("""
SELECT season,
MIN(date) AS first_game_date,
MAX(date) AS last_game_date,
FROM raw.games
WHERE competition_id = 'PL1'
GROUP BY season
ORDER BY season DESC
""").show(max_rows=100)

┌────────┬─────────────────┬────────────────┐
│ season │ first_game_date │ last_game_date │
│ int64  │      date       │      date      │
├────────┼─────────────────┼────────────────┤
│   2025 │ 2025-07-18      │ 2026-05-23     │
│   2024 │ 2024-07-19      │ 2025-05-24     │
└────────┴─────────────────┴────────────────┘



In [22]:
con.sql("""
DESCRIBE raw.appearances
""").show()

┌────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│      column_name       │ column_type │  null   │   key   │ default │  extra  │
│        varchar         │   varchar   │ varchar │ varchar │ varchar │ varchar │
├────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ appearance_id          │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ game_id                │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ player_id              │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ player_club_id         │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ player_current_club_id │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ date                   │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ player_name            │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ competition_id         │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ yellow_cards           │ B

In [35]:
con.sql("""
SELECT 
COUNT(*) AS appearances,
COUNT(DISTINCT game_id) AS total_games,
COUNT(DISTINCT player_id) AS total_players
FROM raw.appearances
WHERE date BETWEEN '2025-07-01' AND '2026-06-30'
AND competition_id = 'NL1'
""").show(max_rows=100)

┌─────────────┬─────────────┬───────────────┐
│ appearances │ total_games │ total_players │
│    int64    │    int64    │     int64     │
├─────────────┼─────────────┼───────────────┤
│        9448 │         306 │           530 │
└─────────────┴─────────────┴───────────────┘



liga NL1, sezon 2025, 306 meczow, 30,9 wystepu na mecz (max 32), 530 zawodnikow, dane kompletne

In [34]:
con.sql("""
    SELECT competition_id, COUNT(*) AS appearances
    FROM raw.appearances
    GROUP BY competition_id
    ORDER BY appearances DESC
""").show(max_rows=100)

┌────────────────┬─────────────┐
│ competition_id │ appearances │
│    varchar     │    int64    │
├────────────────┼─────────────┤
│ IT1            │      155187 │
│ ES1            │      153877 │
│ GB1            │      149751 │
│ FR1            │      143395 │
│ TR1            │      132464 │
│ L1             │      124613 │
│ NL1            │      120019 │
│ PO1            │      119936 │
│ BE1            │      100926 │
│ RU1            │       95397 │
│ GR1            │       87752 │
│ SC1            │       81189 │
│ UKR1           │       76312 │
│ DK1            │       65284 │
│ EL             │       55834 │
│ CL             │       49536 │
│ ELQ            │       19460 │
│ CDR            │       17430 │
│ GRP            │       15505 │
│ RUP            │       14883 │
│ FAC            │       14028 │
│ DFB            │       11370 │
│ CIT            │       11337 │
│ CLQ            │       10580 │
│ NLP            │       10574 │
│ ECLQ           │        9864 │
│ POCP    